### ISSUE TITLE
Metal conv2d fp16 throughput cliffs up to 8x between shapes (M5 Max, MLX 0.32.2)

### ISSUE THREAD

**Issue by dahai80** [dahai80/fusion-mlx#919]
## Summary

Single 3x3 conv2d fp16 microbenchmarks on M5 Max (MLX 0.32.2, idle GPU) show erratic throughput across shapes that matter for MuseTalk UNet/VAE-decode. Same-FLOP shapes differ up to 8x, so SD-VAE decode at 256x256 runs at ~2.6 TFLOPS effective (58ms/frame) instead of the ~10-20 TFLOPS nearby shapes reach.

| shape (NHWC, b=1) | conv2d | GEMM(im2col) |
|---|---|---|
| 256x256, 128->128 | 1.97ms, 9.8 TFLOPS | 14.98ms |
| 256x256, 128->256 | 14.00ms, 2.8 TFLOPS | 17.59ms |
| 128x128, 256->256 | 1.71ms, 11.3 TFLOPS | — |
| 64x64, 512->512 | **8.06ms, 2.4 TFLOPS** | 2.72ms (7.1 TFLOPS) |
| 64x64, 320->320 | 1.82ms, 4.1 TFLOPS | — |
| 32x32, 512->512 | 2.03ms | 1.53ms |
| 32x32, 320->320 | 1.67ms, 1.1 TFLOPS | — |
| 32x32, 1280->1280 | 9.40ms, 3.2 TFLOPS | 4.40ms (6.9) |
| 16x16, 1280->1280 | 3.97ms, 1.9 TFLOPS | 1.17ms (6.5) |

Repro:

```python
import time, mlx.core as mx
x = mx.random.normal((1, 64, 64, 512)).astype(mx.float16)
w = mx.random.normal((512, 3, 3, 512)).astype(mx.float16)
mx.eval(mx.conv2d(x, w, padding=1))
t = time.perf_counter()
for _ in range(20):
    mx.eval(mx.conv2d(x, w, padding=1))
print((time.perf_counter() - t) / 20 * 1000)
```

Notes:
- Batch=4 at 32x32/1280 improves to 6.7 TFLOPS (parallelism-bound kernels).
- An im2col GEMM rewrite wins ~3x at 64x64x512 and 16x16x1280 but loses badly at small-channel shapes — a graph pass could pick per-shape (this would slot into #911).
- Impact: musetalk-mlx hot loop (VAE decode 256x256) is conv-bound at ~58ms/frame; with the good-kernel throughput it would be ~20ms — the difference between the 30FPS PRD target being reachable or not.

## Proposal

- Kernel selection/autotune for the cliff shapes (especially 64x64x512 3x3, 256x256 128->256), or
- A shape-dispatched conv->im2col-GEMM rewrite pass in graph_opt (#911).

**Comment by dahai80** [comment]
**Root cause**: Metal `mx.conv2d` has fp16 throughput cliffs at specific shapes — up to 8× between adjacent configurations (issue table: 64×64 512→512 3×3 at 2.4 TFLOPS vs 7.1 for the same math as im2col GEMM, M5 Max / MLX 0.32.2).

**Fix (v0.10.3, merged to main as 4a36cac2)**: `fusion_mlx/graph_opt/smart_conv.py` — `SmartConv2d` wraps an `nn.Conv2d` and dispatches per call shape between the native conv and an im2col GEMM path (9 shifted slices → `(B·H·W, 9C) @ (9C, O)`).

- Dispatch is a pure shape predicate (Rule 5 — decided in code, no runtime probing), gated on measured wins: reim2col wins at **64×64 512→512 (2×, 0.04→0.02 ms)** and **128×128 256→256 (0.03→0.02 ms)**; native conv wins everywhere else measured (1280-ch, 32×32/16×16, small hw). Thresholds live in `_IM2COL_RULES` and are overridable via `set_im2col_rules()`.
- `apply_smart_conv(root)` wraps supported convs (k=3, s=1, groups=1, same-padding) after weight loading; weights shared by reference; idempotent, no double-wrap.
- fp32 output parity vs `mx.conv2d` is exact on CPU; on GPU Metal GEMM accumulates slightly differently (~1e-3 at the tested shapes), fp16 max diff ≤ 0.014.
- Thresholds regenerate with `scripts/bench_smart_conv.py`.

Opt-in by design — call `apply_smart_conv` explicitly; nothing changes for existing pipelines. Covered by `tests/unit/test_graph_opt_smart_conv.py` (10 tests).

**Comment by dahai80** [comment]
## In-graph regression data (consumer verification, musetalk-mlx)

SmartConv2d parity is clean (UNet cosine 0.999557, VAE decode PSNR 67.43dB — both far above thresholds). But the microbench win does not survive inside the real compiled graph. Same-process A/B on M5 Max / MLX 0.32.2, joint compiled UNet+VAE-decode, batch 2, fp16 (the actual musetalk hot loop):

| variant | p50 per round (2 frames) |
|---|---|
| plain mx.compile | 103.4 ms |
| apply_patterns + apply_smart_conv + mx.compile | **157.0 ms (1.5x slower)** |

This matches an earlier consumer-side experiment: an im2col-GEMM shim was ~3x faster as an isolated microbenchmark but ~7x slower once embedded in the full decode graph.

### Root cause hypothesis

The microbenchmark measures a steady-state loop where the allocator serves the same buffers from cache. In-graph, the im2col path adds `pad + 9 slices + concat` per conv — materializing a (B, H, W, 9*C) tensor (2.25x the input's memory traffic) that competes for the same allocator cache under the compiled graph's allocation pattern. The GEMM's TFLOPS advantage doesn't pay for the extra traffic at these sizes.

### Recommended plan (in order)

1. **Re-calibrate dispatch rules with in-graph measurement.** Bench with the whole decoder under `mx.compile` and toggle `set_im2col_rules` per band, not per-op timing. My data suggests the current `_IM2COL_RULES` bands `(512,512,4096)` / `(256,256,8192)` are net-negative in-graph; the honest default is probably "no bands win" on MLX 0.32.2.
2. **If any band survives: compile the im2col path itself.** Wrapping the pad/slice/concat + matmul in its own `mx.compile` (or ensuring the parent compile fuses the elementwise chain) should cut the materialization cost. Only then does per-shape dispatch make sense.
3. **Long term: fix the Metal conv kernels directly.** The cliffs (#919 report: 64x64x512 3x3 = 2.4 TFLOPS vs 19.8 TFLOPS at 64x64x1280) are a kernel-selection problem inside `mx.conv2d`; a Python-level dispatch is a workaround with overhead, not a fix. Filing upstream at ml-explore/mlx with the shape table would be the durable path.

musetalk-mlx ships `config.SMART_CONV = False` (default) with these numbers recorded in `tests/integration/INTEGRATION_PENDING.md`; happy to re-run the A/B on any candidate change.

**Comment by dahai80** [comment]
Reopening with a follow-up: PR #922 lands real Metal kernel selection for conv2d.

## What landed

`SmartConv2d` now benchmarks two Metal kernels per shape at first call and caches the winner:
- `mx.conv2d` — MLX's tuned Metal conv kernel
- `im2col_conv2d` — im2col gather + MLX matmul Metal kernel

Both candidates are Metal kernels — selection is by measured Metal performance, not a hardcoded rule. Per-machine, per-MLX-version thresholds regenerate with `scripts/bench_smart_conv.py --autotune`.

## Why the original hardcoded rules were wrong

The previous `_IM2COL_RULES` were wrong for M5 Max / MLX 0.32.0:
- `(512,512,4096)` → 64×64 512→512 dispatched to slower im2col (0.675 vs 0.646 ms)
- `(256,256,8192)` → 128×128 256-ch dispatched to slower im2col (0.692 vs 0.472 ms)

They would have slowed the VAE down if wired in. Root cause: `bench_smart_conv.py` had no `mx.eval` inside the timing loop, so it measured graph-build (~0.016 ms) not compute. Fixed.

## Measured

On actual VAE decoder conv shapes: native-sum 6.66 ms → best-per-shape 4.42 ms (1.51× on convs). im2col wins 5/10 bench shapes. VAE decode 256×256: 20.3 → 18.2 ms (-10.5%), no regression at 512×512.

## Baseline finding

VAE decode is already **~20 ms/frame on MLX 0.32.0**. The 58 ms/frame baseline in this issue was measured on MLX 0.32.2 — that throughput cliff does not reproduce on the current MLX (0.32.0). The target is met without new kernel work.

## VAE wiring — default OFF

Wiring into the MuseTalk VAE is opt-in (`FUSION_MUSETALK_SMART_CONV=1`), default OFF. im2col GEMM accumulates in a different order than `mx.conv2d`, so per-conv fp16 drift (~1e-4) compounds through the decoder (mid-block attention + 3 upsample stages) into ~0.4 mean / 5.0 max pixel divergence on a [-1,1] output — visible artifacts. The ~10% speedup is not worth the quality regression on a visual output path.

## Custom Metal kernel — attempted, dropped

A scalar/tiled Metal conv kernel via `mx.fast.metal_kernel` was prototyped as a third candidate but measured 10–30× slower than MLX's tuned conv (which uses `simdgroup_matrix` MMA) everywhere. A competitive version would require a `simdgroup_matrix` GEMM kernel (~150–200 lines, uncertain). Dropped — the autotuner already selects between two real Metal kernels.

## 0.32.2 cliff

Not addressed here (current MLX is 0.32.0, already 20 ms). If 0.32.2 is upgraded to and 58 ms reproduces, file an upstream MLX issue (conv kernel selection regression).

**Comment by dahai80** [comment]
Closed by PR #922 (merged). Autotuning Metal kernel selection landed; VAE decode ~20ms/frame confirmed on MLX 0.32.0 (58ms was 0.32.2-specific). Custom Metal kernel attempt documented as 10-30× slower — dropped.
